janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/addedrelations/SimpleAddedRelations.java

```diff
@@ -14,46 +14,74 @@
 
 package org.janusgraph.graphdb.transaction.addedrelations;
 
-import com.google.common.base.Predicate;
 import org.janusgraph.graphdb.internal.InternalRelation;
 
-import java.util.ArrayList;
+import com.carrotsearch.hppc.ObjectHashSet;
+import com.carrotsearch.hppc.ObjectSet;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import javax.annotation.Nonnull;
+
+import java.util.AbstractCollection;
 import java.util.Collection;
-import java.util.List;
+import java.util.Collections;
+import java.util.Iterator;
 
 /**
  * @author Matthias Broecheler (me@matthiasb.com)
  */
 
-public class SimpleAddedRelations extends ArrayList<InternalRelation> implements AddedRelationsContainer {
```

Comment:
I didn't check the logic much but it looks like it can be simplified to:
```
public class SimpleAddedRelations extends ObjectHashSet<InternalRelation> implements AddedRelationsContainer {
    private static final int INITIAL_ADDED_SIZE = 10;

    public SimpleAddedRelations() {
        super(INITIAL_ADDED_SIZE);
    }

    @Override
    public Iterable<InternalRelation> getView(Predicate<InternalRelation> filter) {
        return Iterables.filter(this::unwrappedIterator, filter);
    }

    @Override
    public Collection<InternalRelation> getAll() {
        return Collections.unmodifiableCollection(new AbstractCollection<InternalRelation>() {
            @Override
            @Nonnull
            public Iterator<InternalRelation> iterator() {
                return SimpleAddedRelations.this.unwrappedIterator();
            }

            @Override
            public int size() {
                return SimpleAddedRelations.this.size();
            }
        });
    }

    private Iterator<InternalRelation> unwrappedIterator() {
        
        return Iterators.transform(iterator(), e -> {
            assert e != null;
            return e.value;
        });
    }
}
```